### PR TITLE
doc(common): documentation improvements

### PR DESCRIPTION
**Pull request by coryan**
The `Status` and `StatusOr` classes are very important and are missing documentation in DocFX. This fleshes out the output a little bit.

The landing page (which I am constantly looking at) had some formatting problems (two `warning` sections back-to-back) and could use some improvements.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11376)

### CODE REVIEW

File: google/cloud/status.h [hunk 1]
```diff
@@ -36,25 +36,162 @@ enum class StatusCode {
...
+  ///
+  /// For operations that may change state within a system, this error may be
+  /// returned even if the operation has completed successfully. For example, a
+  /// successful response from a server could have been delayed long enough for
+  /// the deadline to expire.
   kDeadlineExceeded = 4,
+
+  /// `kNotFound` (gRPC code `NOT_FOUND`) indicates some requested entity (such
+  /// as
+  /// a file or directory) was not found.
```
Comment (devbww) [inline google/cloud/status.h:70]: Line re-flow.

File: google/cloud/status.h [hunk 2]
```diff
@@ -36,25 +36,162 @@ enum class StatusCode {
...
+  /// operation was rejected because the system is not in a state required for
+  /// the operation's execution.
+  ///
+  /// For example, a directory to be deleted may be non-empty, an "rmdir"
+  /// operation is applied to a non-directory, etc.
+  ///
+  /// Some guidelines that may help a service implementer in deciding between
+  /// `kFailedPrecondition`, `kAborted`, and `kUnavailable`:
+  ///
+  /// -# Use `kUnavailable` if the client can retry just the failing call.
```
Comment (devbww) [inline google/cloud/status.h:115]: Q.  Are both the `-` and the `#` significant?

File: google/cloud/status.h [hunk 3]
```diff
@@ -36,25 +36,162 @@ enum class StatusCode {
...
+  ///
+  /// Examples include a per-user quota, or the entire file system being out of
+  /// space.
   kResourceExhausted = 8,
+
+  /// `kFailedPrecondition` (gRPC code `FAILED_PRECONDITION`) indicates that the
+  /// operation was rejected because the system is not in a state required for
+  /// the operation's execution.
+  ///
+  /// For example, a directory to be deleted may be non-empty, an "rmdir"
```
Comment (devbww) [inline google/cloud/status.h:109]: Nit: We say `a "rmdir"` below, which sounds better to my "rrrrrrmdir" pronounciation.

File: google/cloud/status.h [hunk 4]
```diff
@@ -77,8 +214,42 @@ class ErrorInfo {
         domain_(std::move(domain)),
         metadata_(std::move(metadata)) {}
 
+  /**
+   *  The reason of the error.
+   *
+   * This is a constant value that identifies the  proximate cause of the error.
```
Comment (devbww) [inline google/cloud/status.h:220]: s/the<SP><SP>prox/the<SP>prox/

File: google/cloud/status.h [hunk 5]
```diff
@@ -110,23 +281,54 @@ class ErrorInfo {
...
+   * Returns the message associated with the status.
+   *
+   * This is always empty if `code()` is `StatusCode::kOk`.
+   */
   std::string const& message() const;
+
+  /**
+   * Returns the additional error info associated with the status.
+   *
+   * This is always empty if `code()` is `StatusCode::kOk`.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/status.h]: Good point, changed to "default constructed" and documented the constructors.

File: google/cloud/status.h [hunk 6]
```diff
@@ -77,8 +214,42 @@ class ErrorInfo {
...
+   *
+   * For errors generated by the C++ client libraries the domain is
+   * `gcloud-cpp`.
+   */
   std::string const& domain() const { return domain_; }
+
+  /**
+   * Additional structured details about this error.
+   *
+   * Keys should match `/[a-zA-Z0-9-_]/` and be limited to 64 characters in
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/status.h]: Removed the `/`

File: google/cloud/status.h [hunk 7]
```diff
@@ -77,8 +214,42 @@ class ErrorInfo {
...
+   * be at most 63 characters and match a regular expression of
+   * `[A-Z][A-Z0-9_]+[A-Z0-9]`, which represents UPPER_SNAKE_CASE.
+   */
   std::string const& reason() const { return reason_; }
+
+  /**
+   * The logical grouping to which the "reason" belongs.
+   *
+   * The error domain is typically the registered service name of the tool or
+   *  product that generates the error. Example: "pubsub.googleapis.com". If the
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/status.h]: Fixed

File: google/cloud/status.h (hunk 4 above)
Comment (coryan) [reply google/cloud/status.h]: Fixed

File: google/cloud/status.h [hunk 8]
```diff
@@ -36,25 +36,162 @@ enum class StatusCode {
...
+  /// This is typically due to a concurrency issue such as a sequencer check
+  /// failure or a failed transaction.
+  ///
+  /// See the guidelines above for deciding between `kFailedPrecondition`,
+  /// `kAborted`, and `kUnavailable`.
   kAborted = 10,
+
+  /// `kOutOfRange` (gRPC code `OUT_OF_RANGE`) indicates the operation was
+  /// attempted past the valid range, such as seeking or reading past an
+  /// end-of-file.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/status.h:136]: Ack. I am trying to follow the documentation from Abseil and gRPC.  Getting creative (even if it improves correctness) seems risky.

File: google/cloud/status.h (hunk 3 above)
Comment (coryan) [reply google/cloud/status.h]: Fixed

File: google/cloud/status.h (hunk 2 above)
Comment (coryan) [reply google/cloud/status.h]: They are supposed to create numbered items. Replaced with `1.`

File: google/cloud/status.h (hunk 1 above)
Comment (coryan) [reply google/cloud/status.h]: Done.

File: google/cloud/status.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/status.h:136]: Ack.

**Pull request merged**